### PR TITLE
[IOTDB-2453] Remove all not necessarily lock in compaction process

### PR DESCRIPTION
**Pull request by THUMarkLau**

### CODE REVIEW

File: server/src/main/java/org/apache/iotdb/db/engine/compaction/cross/rewrite/task/RewriteCrossSpaceCompactionTask.java [hunk 1]
```diff
@@ -218,11 +193,11 @@ private boolean addReadLock(List<TsFileResource> tsFileResourceList) {
       if (tsFileResource.isCompacting()
           || !tsFileResource.isClosed()
           || !tsFileResource.getTsFile().exists()
-          || tsFileResource.isDeleted()) {
+          || tsFileResource.isDeleted()
+          || !tsFileResource.tryReadLock()) {
```
Comment (qiaojialin) [inline server/src/main/java/org/apache/iotdb/db/engine/compaction/cross/rewrite/task/RewriteCrossSpaceCompactionTask.java:197]:
```suggestion
          || !tsFileResource.readLock()) {
```

File: server/src/main/java/org/apache/iotdb/db/engine/storagegroup/TsFileManager.java [hunk 2]
```diff
@@ -71,6 +71,8 @@ public TsFileManager(
     try {
       List<TsFileResource> allResources = new ArrayList<>();
       Map<Long, TsFileResourceList> chosenMap = sequence ? sequenceFiles : unsequenceFiles;
+      // the iteration of ConcurrentSkipListMap is not concurrent secure
+      // so we must add read lock here
```
Comment (qiaojialin) [inline server/src/main/java/org/apache/iotdb/db/engine/storagegroup/TsFileManager.java:75]: put to readLock

File: server/src/main/java/org/apache/iotdb/db/engine/storagegroup/TsFileManager.java [hunk 3]
```diff
@@ -98,18 +100,13 @@ public TsFileResourceList getUnsequenceListByTimePartition(long timePartition) {
...
-    try {
-      Map<Long, TsFileResourceList> selectedMap = sequence ? sequenceFiles : unsequenceFiles;
-      for (Map.Entry<Long, TsFileResourceList> entry : selectedMap.entrySet()) {
-        if (entry.getValue().contains(tsFileResource)) {
-          entry.getValue().remove(tsFileResource);
-          TsFileResourceManager.getInstance().removeTsFileResource(tsFileResource);
-          break;
-        }
+    Map<Long, TsFileResourceList> selectedMap = sequence ? sequenceFiles : unsequenceFiles;
+    for (Map.Entry<Long, TsFileResourceList> entry : selectedMap.entrySet()) {
```
Comment (qiaojialin) [inline server/src/main/java/org/apache/iotdb/db/engine/storagegroup/TsFileManager.java:104]: add readLock

File: server/src/main/java/org/apache/iotdb/db/engine/storagegroup/TsFileResource.java [hunk 4]
```diff
@@ -507,6 +507,10 @@ public boolean tryWriteLock() {
     return tsFileLock.tryWriteLock();
   }
 
+  public boolean tryReadLock() {
```
Comment (qiaojialin) [inline server/src/main/java/org/apache/iotdb/db/engine/storagegroup/TsFileResource.java:510]: remove

Comment (THUMarkLau) [reply server/src/main/java/org/apache/iotdb/db/engine/storagegroup/TsFileResource.java:510]: resolve

File: server/src/main/java/org/apache/iotdb/db/engine/storagegroup/TsFileManager.java (hunk 3 above)
Comment (THUMarkLau) [reply server/src/main/java/org/apache/iotdb/db/engine/storagegroup/TsFileManager.java:104]: resolve

Comment (THUMarkLau) [reply server/src/main/java/org/apache/iotdb/db/engine/storagegroup/TsFileManager.java:104]: resolve

File: server/src/main/java/org/apache/iotdb/db/engine/storagegroup/TsFileManager.java (hunk 2 above)
Comment (THUMarkLau) [reply server/src/main/java/org/apache/iotdb/db/engine/storagegroup/TsFileManager.java:75]: resolve

File: server/src/main/java/org/apache/iotdb/db/engine/compaction/cross/rewrite/task/RewriteCrossSpaceCompactionTask.java (hunk 1 above)
Comment (THUMarkLau) [reply server/src/main/java/org/apache/iotdb/db/engine/compaction/cross/rewrite/task/RewriteCrossSpaceCompactionTask.java:197]: resolve

File: server/src/main/java/org/apache/iotdb/db/engine/compaction/CompactionScheduler.java [hunk 5]
```diff
@@ -45,8 +42,6 @@
 
   private static IoTDBConfig config = IoTDBDescriptor.getInstance().getConfig();
   // fullStorageGroupName -> timePartition -> compactionCount
```
[the remark this answers is not in the document]
Comment (THUMarkLau) [reply server/src/main/java/org/apache/iotdb/db/engine/compaction/CompactionScheduler.java:44]: resolve

**Review by qiaojialin — APPROVED** [review]

**Pull request merged**